### ISSUE TITLE
Runtime OTEL Metrics for Server

### ISSUE THREAD

**Issue by nikolaydubina** [EricLBuehler/mistral.rs#831]
Would be nice to have Prometheus metrics exported to OTEL (or at least basic /metrics). Useful for production.

https://crates.io/crates/runtime-otel-metrics

**Comment by nikolaydubina** [comment]
if nobody takes this, I will add this myself in coming days.

**Comment by MicheleCampi** [comment]
Hi! I saw this is still open and took a shot at it.
I put together a working proof-of-concept and wanted to check the approach with you before opening a PR.
Branch: https://github.com/MicheleCampi/mistral.rs/tree/feat/prometheus-metrics
What it does:

adds a GET /metrics endpoint exposing the Prometheus text exposition format
installs a process-wide recorder at router build time
adds an axum middleware recording http_requests_total (counter) and http_request_duration_seconds (latency), labeled by method, path, and status

It builds clean (CPU-only here), clippy -D warnings and the existing tests pass, and I've verified the endpoint live against a running server. The footprint is small — one new module plus a few lines in the router builder.
I deliberately kept the PoC minimal to confirm direction first. A few choices I'd like your input on:

1. Library: I used metrics + metrics-exporter-prometheus, since the crate already depends on tracing and this is the lightest companion for a pull-based /metrics. Issue #830 links opentelemetry-rust for traces — would you prefer I align on the OTEL stack from the start, or keep metrics on the lighter metrics facade and treat OTEL traces separately?
2. Histogram vs summary: the Prometheus exporter renders latency as a summary with quantiles by default. For HTTP request latency, bucketed histograms are often preferred since they aggregate across instances. Happy to configure either way — do you have a preference?
3. Inference-level metrics: the PoC instruments the HTTP layer. The natural next step is engine-level metrics (tokens, in-flight requests, queue depth) closer to what vLLM exposes. I'd scope those in a follow-up rather than this first PR — does that split make sense to you?

If the approach looks right, I'll open a PR. I work async and write more than I talk, so I'm glad to iterate here in writing.

**Comment by MicheleCampi** [comment]
Thanks for merging #2189! As I flagged in my earlier comment, the natural next step here is the engine-level runtime metrics (throughput, in-flight requests, queue depth) — the part of this issue that goes beyond the HTTP layer.

I've since looked at how to wire these in, and the good news is the values already exist: `IntervalLogger` (`mistralrs-core/src/engine/logger.rs`) already tracks `tokens_processed`, `num_running`, and `num_waiting` as atomic counters and logs them each interval. So this is largely about *exposing* what's already measured rather than adding new instrumentation. Concretely:

- `mistralrs_tokens_processed_total` (counter)
- `mistralrs_requests_running` (gauge)
- `mistralrs_requests_waiting` (gauge)

Emitted via the same `metrics` facade #2189 already installs, adding `metrics` as a dependency to `mistralrs-core`, with the emission point in the `IntervalLogger` loop where these are already read. One detail: the existing log line is gated on `total_new_seqs != 0 && tokens_processed != 0`, but I'd emit the gauges unconditionally so an idle server still reports `0`.

Planning to open a follow-up PR for this — does the approach and naming look right to you, or anything you'd adjust before I do?
